CorsixTH/Lua/dialogs/resizables/menu_list_dialogs/custom_campaign.lua

```diff
@@ -33,42 +33,63 @@ local col_scrollbar = {
 local details_width = 280
 
 --! Collect the campaign levels at the provided path
---!param path (str) File system path to search.
---!return (array) The found levels, with some basic information about each level.
-local function createCampaignList(path)
-  local campaigns = {}
-
-  for file in lfs.dir(path) do
-    if file:match("%.campaign$") then
-      local campaign_info, err = TheApp:readCampaignFile(file)
-      if not campaign_info then
-        print(err)
+--!param paths_table (array) File system paths to search.
+--!return (array) The found levels, with some basic information about each level, in alphabetical order.
+local function createCampaignList(paths_table)
+  local campaigns, unique_names, duplicates = {}, {}
+  -- Read all campaign files for uniquely named campaigns
+  local function readCampaignFile(folder, file)
+    local full_path = folder .. file
+    local campaign_info, err = TheApp:readCampaignFile(full_path)
+    if not campaign_info then
+      print(err)
+    else
+      local name = campaign_info.name
+      if unique_names[name] then
+        print("Custom campaign error: duplicate campaign name in file " .. file ..
+            ". Check the folders " .. table.concat(paths_table, ", "))
+        duplicates = true
```

Comment:
Avoid using and setting variables that exist outside the function. It makes understanding and debugging a mess. It also makes it impossible to move this local function out of this code without major headaches and/or re-writing.

In my view, Lua allowing this is a major wrong design decision.  
It may "simplify" small scripts, but at our scale the benefit of less input/output parameters is a non-problem compared to difficulties in understanding code structure, and what variables gets used and assigned where.

Reply:
One way to avoid these problems is to only have outer-level functions. Use a `_` prefix to indicate private functions then.

Reply:
Changed to private class functions, and class variables.